### ISSUE TITLE
mocha framework no longer loads dynamically-written tests

### ISSUE THREAD

**Issue by alecf** [webdriverio/webdriverio#5874]
[//]: # NOTE: This repository only maintains packages that are listed in the [Readme](https://github.com/webdriverio/webdriverio/blob/master/README.md#packages). Please make sure that your issue is directly caused by one of these packages and if not file an issue in the correct 3rd party package repository.

**Environment (please complete the following information):**
 - **WebdriverIO version:** 6.5.0
 - **Mode:** WDIO Test runner
 - **If WDIO Testrunner, running sync/async:** async
 - **Node.js version:** 14.9.0
 - **NPM version:** 6.14.7
 - **Browser name and version:** Chrome 85
 - **Platform name and version:** Linux and OSX
 - **Additional wdio packages used (if applicable):** `@wdio/mocha-framework`

**Config of WebdriverIO**
Our webdriverio is pretty vanilla, not much new

**Describe the bug**
We use mocha-style tests with `describe()` and `it()` blocks to describe the tests. 

We have a few data-driven tests, defined like this:
```typescript
const tests = [{
  name: 'homepage',
  url: '/',
}, {
  name: 'profile page',
  url: '/user/profile',
}]
```

and then we define tests:

```typescript
describe('url tests', () => {
  tests.forEach(({ name, url })) => {
    it(`${name}: ${url}`, async () => {
       await navigateTo(url);
       await takeScreenshot();
    });
  });
});
```

These worked fine upto and including 6.4.0, but starting with 6.4.7 the above produces 0 tests and so these tests simply don't run, and tests all pass because nothing failed. Thanks to dependabot we upgraded to 6.5.0 before we realized what broke, and the same thing occurred

**To Reproduce**

[Include code or an example repository that can easily be set up]

**Expected behavior**
I expect the tests to run!

**Additional context**

We're using applitools to take screenshots, though I doubt that is related to any problems here.

the problem occurs both running the tests on browserstack as well as running them locally on a mac.

**Comment by alecf** [comment]
@diachedelic This is my entire log output:
```
Execution of 1 spec files started at 2020-09-21T17:36:30.930Z

Starting ChromeDriver 85.0.4183.87 (cd6713ebf92fa1cacc0f1a598df280093af0c5d7-refs/branch-heads/4183@{#1689}) on port 9515
Only local connections are allowed.
Please see https://chromedriver.chromium.org/security-considerations for suggestions on keeping ChromeDriver safe.
[1600709791.030][WARNING]: FromSockAddr failed on netmask
ChromeDriver was started successfully.
[0-0] RUNNING in chrome - /integration/specs/urlsTest.spec.ts
[0-0] Logging in at http://app.local.uf-dev.com:8080
[0-0] PASSED in chrome - /integration/specs/urlsTest.spec.ts

Spec Files:	 1 passed, 1 total (100% completed) in 00:00:23 

```

it's like it's not seeing any of the tests - the `before` does run (that's where the "Logging in at..." comes from) but nothing else seems to run

**Comment by alecf** [comment]
oh, I just dug around with `--logLevel info` and found I had a stray call that wasn't inside an describe/it:
```ts
describe('URL Tests', () => {
  before(async () => {... });
  // oops, this is broken:
  browser.setTimeout({ script: 60000 * 10 });
  tests.forEach(({ name, url })) => {
    it(`${name}: ${url}`, async () => {
       await navigateTo(url);
       await takeScreenshot();
    });
  });
});
```

So this is user error, that for some strange reason worked correctly *before* the `@wdio/mocha-framework` upgrade. this was my error!

**Comment by alecf** [comment]
actually, I'm gonna reopen this because the failure mode was pretty bad - an exception was being thrown saying `browser.setTimeout()` didn't exist, but I *only* found that by setting the log level to `info` - this should have failed the whole suite IMO

**Comment by diachedelic** [comment]
I agree that if a spec fails to load, wdio should fail fast and loudly.

**Comment by diachedelic** [comment]
From the [tests](https://github.com/webdriverio/webdriverio/blob/a4a5a46f786f8548361d7f86834b38f89dcb1690/packages/wdio-mocha-framework/tests/adapter.test.js#L383):
```js
test('should not fail on exception', async () => {
        const adapter = adapterFactory({})
        adapter._hasTests = null
        adapter.mocha = {
            loadFilesAsync: jest.fn().mockImplementation(
                () => Promise.reject(new Error('foo'))
            ),
        }
        await adapter._loadFiles({})
        expect(adapter.mocha.loadFilesAsync).toBeCalled()
        expect(adapter._hasTests).toBe(null)
    })
```

It looks fairly intentional...

**Comment by framauro** [comment]
I am also seeing an issue with 6.4.7+ mocha-framework.

In my case, what I see is tests starting _before_ the before() hook is complete... 

The before() hook does completes, but after all the test specs have failed.

I am using before() hook to setup global variables, which are used in dynamic test specs. These globals e.g. 'sites' are not set by the time the spec is executed.  i.e. before() no longer appears to run _before_ the specs! Until 6.4.0 this has worked fine.

```
2020-09-25T17:09:16.605Z INFO @wdio/cli:launcher: Run onWorkerStart hook
2020-09-25T17:09:16.605Z INFO @wdio/cli:launcher: Run onWorkerStart hook
2020-09-25T17:09:16.605Z INFO @wdio/cli:launcher: Run onWorkerStart hook
2020-09-25T17:09:16.605Z INFO @wdio/cli:launcher: Run onWorkerStart hook
2020-09-25T17:09:16.607Z INFO @wdio/local-runner: Start worker 0-0 with arg: test/config/wdio.conf.js
2020-09-25T17:09:16.612Z INFO @wdio/local-runner: Start worker 0-1 with arg: test/config/wdio.conf.js
2020-09-25T17:09:16.616Z INFO @wdio/local-runner: Start worker 1-0 with arg: test/config/wdio.conf.js
2020-09-25T17:09:16.620Z INFO @wdio/local-runner: Start worker 1-1 with arg: test/config/wdio.conf.js
[0-0] 2020-09-25T17:09:16.928Z INFO @wdio/local-runner: Run worker command: run
[0-1] 2020-09-25T17:09:16.931Z INFO @wdio/local-runner: Run worker command: run
[1-0] 2020-09-25T17:09:16.932Z INFO @wdio/local-runner: Run worker command: run
[1-1] 2020-09-25T17:09:16.939Z INFO @wdio/local-runner: Run worker command: run
[0-0] 2020-09-25T17:09:17.246Z INFO webdriverio: Initiate new session using the ./protocol-stub protocol
[1-0] 2020-09-25T17:09:17.246Z INFO webdriverio: Initiate new session using the ./protocol-stub protocol
[0-1] 2020-09-25T17:09:17.247Z INFO webdriverio: Initiate new session using the ./protocol-stub protocol
[1-1] 2020-09-25T17:09:17.254Z INFO webdriverio: Initiate new session using the ./protocol-stub protocol
[0-0] 2020-09-25T17:09:17.419Z WARN @wdio/mocha-framework: Unable to load spec files quite likely because they rely on `browser` object that is not fully initialised.
`browser` object has only `capabilities` and some flags like `isMobile`.
Helper files that use other `browser` commands have to be moved to `before` hook.
Spec file(s): /Users/*****/IdeaProjects/jumpmangaming/wdio-joinnow-check/test/specs/joinnow.check.spec.js
 Error:  ReferenceError: sites is not defined
    at Object.<anonymous> (/Users/*****/IdeaProjects/jumpmangaming/wdio-joinnow-check/test/specs/joinnow.check.spec.js:1:1)
    at Module._compile (internal/modules/cjs/loader.js:1137:30)
    at Module._compile (/Users/*****/IdeaProjects/jumpmangaming/wdio-joinnow-check/node_modules/pirates/lib/index.js:99:24)
    at Module._extensions..js (internal/modules/cjs/loader.js:1157:10)
    at Object.newLoader [as .js] (/Users/*****/IdeaProjects/jumpmangaming/wdio-joinnow-check/node_modules/pirates/lib/index.js:104:7)
    at Module.load (internal/modules/cjs/loader.js:985:32)
    at Function.Module._load (internal/modules/cjs/loader.js:878:14)
    at Module.require (internal/modules/cjs/loader.js:1025:19)
    at require (internal/modules/cjs/helpers.js:72:18)
    at Object.exports.requireOrImport (/Users/*****/IdeaProjects/jumpmangaming/wdio-joinnow-check/node_modules/mocha/lib/esm-utils.js:20:12)
```

**Comment by framauro** [comment]
I am also seeing an issue with 6.4.7+ mocha-framework.

In my case, what I see is tests starting _before_ the before() hook is complete... 

The before() hook does completes, but after all the test specs have failed.

I am using before() hook to setup global variables, which are used in dynamic test specs. These globals e.g. 'sites' are not set by the time the spec is executed.  i.e. before() no longer appears to run _before_ the specs! Until 6.4.0 this has worked fine.

```
2020-09-25T17:09:16.605Z INFO @wdio/cli:launcher: Run onWorkerStart hook
2020-09-25T17:09:16.605Z INFO @wdio/cli:launcher: Run onWorkerStart hook
2020-09-25T17:09:16.605Z INFO @wdio/cli:launcher: Run onWorkerStart hook
2020-09-25T17:09:16.605Z INFO @wdio/cli:launcher: Run onWorkerStart hook
2020-09-25T17:09:16.607Z INFO @wdio/local-runner: Start worker 0-0 with arg: test/config/wdio.conf.js
2020-09-25T17:09:16.612Z INFO @wdio/local-runner: Start worker 0-1 with arg: test/config/wdio.conf.js
2020-09-25T17:09:16.616Z INFO @wdio/local-runner: Start worker 1-0 with arg: test/config/wdio.conf.js
2020-09-25T17:09:16.620Z INFO @wdio/local-runner: Start worker 1-1 with arg: test/config/wdio.conf.js
[0-0] 2020-09-25T17:09:16.928Z INFO @wdio/local-runner: Run worker command: run
[0-1] 2020-09-25T17:09:16.931Z INFO @wdio/local-runner: Run worker command: run
[1-0] 2020-09-25T17:09:16.932Z INFO @wdio/local-runner: Run worker command: run
[1-1] 2020-09-25T17:09:16.939Z INFO @wdio/local-runner: Run worker command: run
[0-0] 2020-09-25T17:09:17.246Z INFO webdriverio: Initiate new session using the ./protocol-stub protocol
[1-0] 2020-09-25T17:09:17.246Z INFO webdriverio: Initiate new session using the ./protocol-stub protocol
[0-1] 2020-09-25T17:09:17.247Z INFO webdriverio: Initiate new session using the ./protocol-stub protocol
[1-1] 2020-09-25T17:09:17.254Z INFO webdriverio: Initiate new session using the ./protocol-stub protocol
[0-0] 2020-09-25T17:09:17.419Z WARN @wdio/mocha-framework: Unable to load spec files quite likely because they rely on `browser` object that is not fully initialised.
`browser` object has only `capabilities` and some flags like `isMobile`.
Helper files that use other `browser` commands have to be moved to `before` hook.
Spec file(s): /Users/*****/IdeaProjects/****/wdio-joinnow-check/test/specs/joinnow.check.spec.js
 Error:  ReferenceError: sites is not defined
    at Object.<anonymous> (/Users/*****/IdeaProjects/****/wdio-joinnow-check/test/specs/joinnow.check.spec.js:1:1)
    at Module._compile (internal/modules/cjs/loader.js:1137:30)
    at Module._compile (/Users/*****/IdeaProjects/****/wdio-joinnow-check/node_modules/pirates/lib/index.js:99:24)
    at Module._extensions..js (internal/modules/cjs/loader.js:1157:10)
    at Object.newLoader [as .js] (/Users/*****/IdeaProjects/****/wdio-joinnow-check/node_modules/pirates/lib/index.js:104:7)
    at Module.load (internal/modules/cjs/loader.js:985:32)
    at Function.Module._load (internal/modules/cjs/loader.js:878:14)
    at Module.require (internal/modules/cjs/loader.js:1025:19)
    at require (internal/modules/cjs/helpers.js:72:18)
    at Object.exports.requireOrImport (/Users/*****/IdeaProjects/****/wdio-joinnow-check/node_modules/mocha/lib/esm-utils.js:20:12)
```

**Comment by mgrybyk** [comment]
Dont use browser object until the session is started as mentioned in the error

**Comment by mgrybyk** [comment]
Anyway, absolutely agree with https://github.com/webdriverio/webdriverio/issues/5874#issuecomment-696448659

**Comment by diachedelic** [comment]
@framauro it could be that Mocha applies stricter rules regarding globals when using `import()` rather than `require()` (a change which was introduced by #5809). Is there any reason you need to use global variables, and cannot just import `sites` as a module?

**Comment by christian-bromann** [comment]
I fixed this in #5910
